Addressing #6082 and #7766: warning when overriding notation format + new notion of format associated to a given interpretation

herbelin opened the pull request:
Closes #6082 
Closes #7766
~~Closes 9741~~ (moved to #11590)

Upstream overlays to be merged synchronously:  [Coq-Equations #254](https://github.com/mattam82/Coq-Equations/pull/254), [QuickChick #183](https://github.com/QuickChick/QuickChick/pull/183).

**Kind:** bug fix / enhancement

We propose a solution to fix #6082 and #7766 by doing the following changes:
- We introduce a notion of format specifically attached to a given interpretation distinct from the former notion of format (generically) attached to a notation.
- We warn before overriding a generic format, or a specific format.

The effect of commands is accordingly the following:
- `Reserved Notation` attaches a format generically.
- `Notation` attaches the associated or inferred format specifically to the given interpretation
-  if `Notation` comes with some parsing data (such as `at level`), it additionally tries to attach the associated or inferred format generically to the notation (seen as an uninterpreted string) [see design question below (**)].

To do this, we add a new field to `CNotation`/`CPatNotation` indicating that the corresponding notation comes from a specific interpretation. This field is `None` when coming from parsing and `Some` when intented to be used by the printer.

The description of commits is the following:
- The first commit is a preliminary renaming work.
- The main changes of the PR are in the second commit, in `constrextern.ml`, `metasyntax.ml`, `ppextend.ml`, `ppconstr.ml`. Changes in the other files are adaptation to the main changes.
- The third commit is a cleaning work.
- An extra fourth commit experiments the following semantic change: if a notation is independenty being given an only-parsing rule and an only-printing rule, it is an error if they do not have the same level and sublevels.

Side question: are extra `format` rules used in the large? Is `Ppextend.find_notation_extra_printing_rules` known to be used in external contributions? (I also made a distinction between extra rules attached generically and extra rules attached specifically though.)

About the design questions (**), what to do in a situation such as:
```
Notation "x %% y" := (x+y) (at level 47, format "x   %%   y") : foo_scope.
Notation "x %% y" := (x*y) : bar_scope.
```
Should the second notation inherits the (specific) format of the first declaration of a notation `" x %% y"` or uses instead the default format built for the notation? Otherwise said, should the format given the first time a given notation is declared (i.e. w/o a prior `Reserved Notation`) be considered (also) as the default generic format to be used by all further interpretations of the notation or not? I implemented a yes to the question but using the default one may be better, no? [Updated: I backtracked on this choice, see this [comment](https://github.com/coq/coq/pull/10832#issuecomment-553119939), but a use case for the other answer in Iris, [below](https://github.com/coq/coq/pull/10832#issuecomment-585739953), makes me hesitating.]

- [X] Added / updated test-suite
- [X] Documentation of the distinction between specific and generic format
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

herbelin commented:
For the record:
- Equations and Quickchick require an ML overlay for `CNotation`
- Iris: seems to fix an instance of #6082 on the Iris test-suite
- Bedrock2 printing in `Bytedump.v` can now probably be made less hackish (see #9744).

maximedenes commented on interp/constrextern.ml:
```diff
@@ -107,18 +107,19 @@ let deactivate_notation nr =
      inactive_notations_table := IRuleSet.add nr !inactive_notations_table
-  | NotationRule (scopt, ntn) ->
-     match availability_of_notation (scopt, ntn) (scopt, []) with
+  | NotationRule (which, ntn) ->
```
The name `which` does not seem very informative.

maximedenes replied:
(idem for other occurrences below)

herbelin replied:
Yes, `which` is not a very discriminating word. I thought at `kind` at some time but it would not be better. Maybe `specif` since it is in the type `notation_specificity`, but at the same time, `specificity` is not very "specific" here. The idea is to tell whether it is in some scope or lonely. So, maybe `isinscope`, or `scopekind`? We could also keep the original `scopt` and rename `notation_specificity` into `notation_optionally_in_scope`, or `notation_with_scope`. I'm missing imagination. Maybe would I then choose `scopt` and `notation_with_scope` if noone has better to propose.

maximedenes commented on interp/constrexpr.ml:
```diff
@@ -19,11 +19,17 @@ type universe_decl_expr = (lident list, Glob_term.glob_constraint list) UState.g
 type ident_decl = lident * universe_decl_expr option
 type name_decl = lname * universe_decl_expr option
 
+type notation_specificity = LastLonelyNotation | NotationInScope of string
```
What does `specificity` mean, in this context?

herbelin replied:
See above. I'm not particularly satisfied with the use of the word `specificity` here. In any case, thanks a lot for reviewing.

herbelin commented:
> I'd like to understand if printing rules are considered as part of the "syntactic" or the "interpretation" part of a notation. 

With this PR, printing rules are related to both the "syntactic" or the "interpretation" part of a notation: the generic ones are attached to the syntax and the specific ones to the interpretation.

> If it is the former, then I'm not sure why `add_notation_interpretation_core` deals with them.

The reason why the function `add_notation_interpretation_core` redeclare the syntactic part is a question of import (traced back to c32fcecf66c). If I define in file `B.v` a new interpretation for a notation, say, `"_ *** _"` whose original syntax was defined in file `A.v` which `B` imported (but w/o re-exporting `A`), then, if I import `B` but w/o importing `A`, I still have to ensure that the parsing/printing rules for `"_ *** _ "` is declared. (This constraint is independent of the PR.)

>  If the latter, it means that `Reserved Notation` now declares a syntactic and an interpretation part, and the separation should be made clear in the code. (maybe it is, but I couldn't see it)

`Reserved Notation` declares only a syntactic part.

Does it answer your question? Do you see ways to make the code clearer?

herbelin commented:
I changed `notation_specificity` to `notation_with_optional_scope` and `which` to `inscope`.

If no other comments, I shall proceed with the overlays, starting with Equations.

ppedrot commented:
I am postponing this. If this is deemed to be considered a fix only, it might get in between the merge and the beta though.

herbelin commented:
Update of the PR:
- I resolved initial question (**) by not declaring the specific format given to the first occurrence of a (non-reserved) notation to be also the generic one; instead, in this case, the generic format is the one canonically derived from the notation string.
- I made overlays
- I made a short addition to the documentation (maybe more could be made, please tell)

So, from my point of view, the PR is ready. @maximedenes, since you started to review, would like to be the assignee (note my changes above in response to your comments)?

herbelin commented:
@maximedenes: are you still interested in reviewing and assigning this PR? Is there something that can be done to help you?

ejgallego commented:
Build is broken @herbelin

herbelin commented:
@JasonGross:

This PR gives a warning `Lonely notation "rew dependent _ in _" was already defined with a different format` in `Specif.v`:
```
Local Notation "'rew' 'dependent' H 'in' H'"   
  := (match H with                      
      | eq_refl => H'      
      end)           
       (at level 10, H' at level 10,   
        format "'[' 'rew'  'dependent'  '/    ' H  in  '/' H' ']'").   
```
while, in `Logic.v`, it was previously:
```
Notation "'rew' 'dependent' H 'in' H'"  
  := (match H with      
      | eq_refl => H'    
      end)   
       (at level 10, H' at level 10,   
        format "'[' 'rew'  'dependent'  H  in  '/' H' ']'"). 
```
What would you prefer, that we choose twice the same format, and if yes which one? The second one, which seems more elaborated? Or, is this different format intended and we add a line to silence the warning?

JasonGross commented:
The difference is not intended.  Feel free to go with whichever one seems best to you, and drop the local Notation altogether

herbelin commented:
> The difference is not intended. Feel free to go with whichever one seems best to you, and drop the local Notation altogether

OK, I generalized the allowance for a break before the « `H` ».

herbelin commented:
The failure in perennial seems expected: there is a mismatch in the levels between an only-printing rule and a both-parsing-printing rule. I guess the second should be only parsing, and it would be better that they agree anyway.

The failure in Iris is related to the design choice (**). Iris uses the notation `∃ x .. y , P` from `Utf8_core.v`, but `Utf8_core.v` does not reserve the notation, thus, its format is not (anymore) inherited in iris. Should I backtrack on how I resolved `(**)`, or should we reserve the format of `∃ x .. y , P` in `Utf8_core.v` so that it can be inherited. I'm a bit hesitant. Cc @ralfjung.

Failure in Equations is due to overlay needing rebasing. In Quickchick to a temporary worry with #11235.

RalfJung commented:
Cc @robbertkrebbers  see Hugo's question above.

> The failure in Iris is related to the design choice (**). Iris uses the notation ∃ x .. y , P from Utf8_core.v, but Utf8_core.v does not reserve the notation, thus, its format is not (anymore) inherited in iris.

So you changed (**) from what it says in the OP? There it still says things *do* get inherited.

> * Reserved Notation attaches a format generically.
> * Notation attaches the associated or inferred format specifically to the given interpretation

I am confused... how is it even possible for notation to have a different format specifically than generically? I thought this would be like levels and all the other notation annotations, where all notations across all scopes have to share these settings.

herbelin commented:
> So you changed (**) from what it says in the OP? There it still says things do get inherited.

Yes, I changed it at the time of this [comment](https://github.com/coq/coq/pull/10832#issuecomment-553119939). But now I'm confused again about which policy would be the most expected.

> how is it even possible for notation to have a different format specifically than generically?

It means that the format can be different depending on scope (different here means different spacing or formatting boxes but same content, e.g. because the use of list in scome scope migh be expected to be encapsulated in a vertical box, but other uses would not).

Incidentally, I wonder if Iris is also using multi-printing rules (in the same scope) like perennial is using, as shown in #11591.

RalfJung commented:
> Yes, I changed it at the time of this comment. But now I'm confused again about which policy would be the most expected.

It would be helpful to update this in the OP, to keep it a useful summary of the current status.

> It means that the format can be different depending on scope (different here means different spacing or formatting boxes but same content, e.g. because the use of list in scome scope migh be expected to be encapsulated in a vertical box, but other uses would not).

So, this is actually implementable? But having the level depend on the scope is not?

> Incidentally, I wonder if Iris is also using multi-printing rules (in the same scope) like perennial is using

I think if we do it is an accident. But I am not sure.

herbelin commented:
> It would be helpful to update this in the OP, to keep it a useful summary of the current status.

You're right. I updated.

> > It means that the format can be different depending on scope
>
> So, this is actually implementable?

For printing, you could somehow do whatever you want as soon as you don't care about reparsing. The tolerance into different spacing is not very strong (the only effect could be that two tokens which should have been disjoint become a different single one because a space is missing between them). The expectation of having the same levels as parsing is for reparsing.

> But having the level depend on the scope is not?

That's not possible for parsing: you need to know the levels in advance to disambiguate parsing (i.e. to turn a list into a tree) and you know the scope only after the fact (either after parsing the expression if a `%scope`, or even at interpretation time if the scope is attached to a constant). Would the syntax for `%scope` have been prefixed (as e.g. `scope%(...)`), that would have been different (but then custom entries come the rescue).

RalfJung commented:
When I reported https://github.com/coq/coq/issues/7766, I was under the expectation that just like all variants of one notation in different scopes have to have the same level, they also all have to have the same formatting. I'd still be fine with that, if it helps simplify the implementation.

But when that restriction is lifted... we'd still like the Iris existential quantifier to be formatted the same way as the one in the standard library, if possible. So ideally there would be a way for us to *not* have to repeat the format string, and just inhehit it like we inherit levels. I think this corresponds to either your original plan for (**), or else add an explicit reservation in stdlib.

ejgallego commented:
This PR seems very close to converge, albeit we may want the discussion about inheritance to develop a bit more, right?

Anyways I'm setting a merge target for beginning of next, please comment if you would like to think more about this.

herbelin commented:
I just experimented going back to the initial proposal, which is that the first declaration of a notation (w/o using `Reserved`) sets the generic format. There is difference with reserved notations though: it is not a warning if another interpretation of the same notation syntax uses a different format, while it is a warning if this syntax was reserved with a format.

In the meantime, I also introduced yesterday a check that no `format` is given for `only parsing` notations. This shows that such combinations occur in math-comp (`n %:Z` in `ssrint.v`), `bedrock2` (file `NotationsCustomEntry.v`) and fiat-parsers (file `BuildADT.v`).

Even if it is does not make sense per se, combining `only parsing` and `format` is legitimate while in a development process, so I now moved the error into a warning.

ejgallego commented:
Based on the discussion will merge once the CI passes if no further comment.

herbelin commented:
I suspect the fiat-crypto failure to be related to #11530: it write:
```
      = bedrock_func_body:("x0" =   
                           (expr.load access_size.word         
                              (expr.op bopname.add (expr.var "y0")  
                                 (expr.literal 0)));  
```
instead of
```
      = bedrock_func_body:((((("x0" =     
                               (load( constr:(expr.var "y0") +  
                                      constr:(expr.literal 0))));   
```
I'll wait some time before relaunching.
